From 02055de1ff1b15c7c12bff0457bae67c3db97f0a Mon Sep 17 00:00:00 2001
From: "Richard W.M. Jones" <rjones@redhat.com>
Date: Mon, 29 Jan 2024 18:20:46 +0000
Subject: [PATCH] block/blkio: Don't assume size_t is 64 bit
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

With GCC 14 the code failed to compile on i686 (and was wrong for any
version of GCC):

../block/blkio.c: In function ‘blkio_file_open’:
../block/blkio.c:857:28: error: passing argument 3 of ‘blkio_get_uint64’ from incompatible pointer type [-Wincompatible-pointer-types]
  857 |                            &s->mem_region_alignment);
      |                            ^~~~~~~~~~~~~~~~~~~~~~~~
      |                            |
      |                            size_t * {aka unsigned int *}
In file included from ../block/blkio.c:12:
/usr/include/blkio.h:49:67: note: expected ‘uint64_t *’ {aka ‘long long unsigned int *’} but argument is of type ‘size_t *’ {aka ‘unsigned int *’}
   49 | int blkio_get_uint64(struct blkio *b, const char *name, uint64_t *value);
      |                                                         ~~~~~~~~~~^~~~~

Signed-off-by: Richard W.M. Jones <rjones@redhat.com>
---
 block/blkio.c | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/block/blkio.c b/block/blkio.c
index 0a0a6c0f5fd..52d78935147 100644
--- a/block/blkio.c
+++ b/block/blkio.c
@@ -794,6 +794,7 @@ static int blkio_file_open(BlockDriverState *bs, QDict *options, int flags,
     const char *blkio_driver = bs->drv->protocol_name;
     BDRVBlkioState *s = bs->opaque;
     int ret;
+    uint64_t val;
 
     ret = blkio_create(blkio_driver, &s->blkio);
     if (ret < 0) {
@@ -854,7 +855,7 @@ static int blkio_file_open(BlockDriverState *bs, QDict *options, int flags,
 
     ret = blkio_get_uint64(s->blkio,
                            "mem-region-alignment",
-                           &s->mem_region_alignment);
+                           &val);
     if (ret < 0) {
         error_setg_errno(errp, -ret,
                          "failed to get mem-region-alignment: %s",
@@ -862,6 +863,15 @@ static int blkio_file_open(BlockDriverState *bs, QDict *options, int flags,
         blkio_destroy(&s->blkio);
         return ret;
     }
+#if HOST_LONG_BITS == 32
+    if (val > SIZE_MAX) {
+        error_setg_errno(errp, ERANGE,
+                         "mem-region-alignment too large for size_t");
+        blkio_destroy(&s->blkio);
+        return -ERANGE;
+    }
+#endif
+    s->mem_region_alignment = (size_t)val;
 
     ret = blkio_get_bool(s->blkio,
                          "may-pin-mem-regions",
-- 
2.43.0

